### PR TITLE
[rnncell] Implement rnncell

### PR DESCRIPTION
**Pull request by lhs8928**
Commit 1:     [Bugfix] Regenerate lstmcell unittest
    
     - Set lstm bias_ih to unused in init to fix the bug
     - Regenerate the lstmcell unittest
     - Fix typo
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 2:     [rnncell] Implement rnncell
    
     - Implement rnncell which is only run on 1 timestep
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

Commit 3:     [unittest] Implement rnncell unittest
    
     - Generate rnn, rnncell layer unittest
     - Generate model unittest which is composed of rnncell
     - Verified with multi timestep/stacked/loop_unrolling situation
    
    Self evaluation:
    
    Build test: [X]Passed [ ]Failed [ ]Skipped
    Run test: [X]Passed [ ]Failed [ ]Skipped
    
    Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

File: nntrainer/layers/rnncell.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+  rnncell_props(props::Unit(), props::HiddenStateActivation(), props::DropOutRate(),
+                props::MaxTimestep(), props::Timestep()),
+  wt_idx({0}),
+  acti_func(ActivationType::ACT_NONE, true),
+  epsilon(1e-3) {}
+
+void RNNCellLayer::finalize(InitLayerContext &context) {
+  nntrainer::WeightRegularizer &weight_regularizer =
+    std::get<props::WeightRegularizer>(*layer_impl_props);
+  float &weight_regularizer_constant =
```
Comment (zhoonit) [inline nntrainer/layers/rnncell.cpp:43]: fyi) for the primitives like float, int, pointers or enum like (Tensor::Initializer), it is preferred to return by value if we are just reading the value, + if we are reading the value, it'd better be const reference.

File: nntrainer/layers/rnncell.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+  context.setOutputDimensions({output_dim});
+
+  // weight_xh_dim : [1, 1, input_size, unit]
+  TensorDim weight_xh_dim({feature_size, unit});
+  // weight_hh_dim : [1, 1, unit, unit]
+  TensorDim weight_hh_dim({unit, unit});
+  // bias_h_dim : [1, 1, 1, unit]
+  TensorDim bias_h_dim({unit});
+
+  // weight_initializer can be set sepeartely. weight_xh initializer,
```
Comment (zhoonit) [inline nntrainer/layers/rnncell.cpp:86]: may be typo? `sepeartely`

File: nntrainer/layers/rnncell.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+  // test will fail. Becuase it modifies the data during gradient calculation
+  // TODO : We could control with something like #define test to save memory
+  TensorDim dim(batch_size * max_timestep, 1, 1, unit);
+  wt_idx[RNNCellParams::hidden_state] = context.requestTensor(
+    dim, context.getName() + ":hidden_state", Tensor::Initializer::NONE, true,
+    TensorLifespan::ITERATION_LIFESPAN);
+
+  if (hidden_state_activation_type.get() == ActivationType::ACT_NONE) {
+    hidden_state_activation_type.set(ActivationType::ACT_TANH);
+  }
```
Comment (zhoonit) [inline nntrainer/layers/rnncell.cpp:109]: looks like it's better that hidden_state_activation_type default is set to ACT_TANH instead

**Review by zhoonit — APPROVED** [review]
Minor comment, overall LGTM! nice job :)

File: nntrainer/layers/rnncell.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+RNNCellLayer::RNNCellLayer() :
+  LayerImpl(),
+  rnncell_props(props::Unit(), props::HiddenStateActivation(), props::DropOutRate(),
+                props::MaxTimestep(), props::Timestep()),
+  wt_idx({0}),
+  acti_func(ActivationType::ACT_NONE, true),
+  epsilon(1e-3) {}
+
+void RNNCellLayer::finalize(InitLayerContext &context) {
```
Comment (zhoonit) [inline nntrainer/layers/rnncell.cpp:40]: We need a check if num inputs == 1, within this function

File: nntrainer/compiler/recurrent_realizer.cpp [hunk 5]
```diff
@@ -128,7 +129,8 @@ static void propagateTimestep(LayerNode *node, unsigned int time_step,
   /** @todo add an interface to check if a layer supports a property */
   auto is_recurrent_type = [](LayerNode *node) {
     return node->getType() == LSTMLayer::type ||
-           node->getType() == LSTMCellLayer::type;
+           node->getType() == LSTMCellLayer::type ||
+           node->getType() == RNNCellLayer::type;
```
Comment (jijoongmoon) [inline nntrainer/compiler/recurrent_realizer.cpp:133]: It isn't just for the cell implementation. if it is, let's add others. GRU as well.

File: nntrainer/layers/rnncell.cpp [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+                          weight_regularizer_constant, "weight_hh", true);
+  wt_idx[RNNCellParams::bias_h] =
+    context.requestWeight(bias_h_dim, bias_initializer, WeightRegularizer::NONE,
+                          1.0f, "bias_h", true);
+
+  // We do not need this if we reuse net_hidden[0]. But if we do, then the unit
+  // test will fail. Becuase it modifies the data during gradient calculation
+  // TODO : We could control with something like #define test to save memory
+  TensorDim dim(batch_size * max_timestep, 1, 1, unit);
+  wt_idx[RNNCellParams::hidden_state] =
```
Comment (jijoongmoon) [inline nntrainer/layers/rnncell.cpp:104]: the demension for hidden includes time step and input dimension does not.

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/compiler/recurrent_realizer.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/compiler/recurrent_realizer.cpp:133]: This is only for the cell implementation. LSTMLayer supports unoptimized LSTMcell implementation, but RNNLayer and GRULayer do not. GRUCell will be added here once its ready.

File: nntrainer/layers/rnncell.cpp (hunk 1 above)
Comment (lhs8928) [reply nntrainer/layers/rnncell.cpp]: Done :)

File: nntrainer/layers/rnncell.cpp (hunk 2 above)
Comment (lhs8928) [reply nntrainer/layers/rnncell.cpp]: Fixed

File: nntrainer/layers/rnncell.cpp (hunk 4 above)
Comment (lhs8928) [reply nntrainer/layers/rnncell.cpp:40]: Already existed :) 
You can check it at line 56

File: nntrainer/layers/rnncell.cpp (hunk 6 above)
Comment (lhs8928) [reply nntrainer/layers/rnncell.cpp:104]: Yes the input dimension doesn't include time step but to run forwarding and backwarding the previous hidden state is needed.
So hidden state dimension includes max_timestep to store the result at each timestep.

File: nntrainer/layers/rnncell.cpp (hunk 3 above)
Comment (lhs8928) [reply nntrainer/layers/rnncell.cpp:110]: The default hidden state activation type of RNN, LSTM, GRU which we've implemented is TANH but I'm not sure for other layers.

Comment (zhoonit) [reply nntrainer/layers/rnncell.cpp:110]: Yeah, how about at the constructor we initialize with the default value?

File: nntrainer/layers/rnncell.cpp (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/rnncell.cpp:40]: Right, wonder how I missed it :p

**Review by kparichay — APPROVED** [review]
LGTM

File: test/input_gen/genModelsRecurrent_v2.py [hunk 7]
```diff
@@ -42,12 +41,15 @@ def __init__(self, unroll_for=2, num_lstm=1):
         # self.lstm.weight_hh.data.fill_(1.0)
         # self.lstm.weight_ih.data.fill_(1.0)
         # self.lstm.bias_hh.data.fill_(1.0)
+        for lstm in self.lstms:
+            lstm.bias_ih.data.fill_(0.0)
+            lstm.bias_ih.requires_grad=False
```
Comment (zhoonit) [inline test/input_gen/genModelsRecurrent_v2.py:80]: here

**Pull request merged**